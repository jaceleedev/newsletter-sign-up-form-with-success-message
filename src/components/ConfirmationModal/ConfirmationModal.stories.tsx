import type { Meta, StoryObj } from '@storybook/react';
import ConfirmationModal from './ConfirmationModal';

const meta = {
  component: ConfirmationModal,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    email: {
      description: 'Email address to be displayed in the confirmation message',
      control: 'text',
    },
  },
} satisfies Meta<typeof ConfirmationModal>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    email: 'ash@loremcompany.com',
  },
};
