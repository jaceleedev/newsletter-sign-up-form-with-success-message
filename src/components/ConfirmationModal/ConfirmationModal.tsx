'use client';

import Image from 'next/image';
import Button from '../Button';
import { useState } from 'react';

interface ConfirmationModalProps {
  email: string;
}

function ConfirmationModal({ email }: Readonly<ConfirmationModalProps>) {
  const [isVisible, setIsVisible] = useState(true);

  function handleClick() {
    setIsVisible((prev) => !prev);
  }

  if (!isVisible) {
    return null;
  }

  return (
    <div className="max-w-[375px] h-[812px] px-6 pt-[149px] pb-10 bg-white lg:max-w-[504px] lg:h-[520px] lg:px-16 lg:pt-12 lg:pb-16 lg:rounded-[36px] lg:shadow-[0_15px_60px_0_rgba(0,0,0,0.25)]">
      <div className="mb-[263px] lg:mb-10">
        <Image
          src="/images/icon-success.svg"
          alt=""
          width={64}
          height={64}
          priority
          className="mb-10"
        />
        <h1 className="text-heading mb-6">Thanks for subscribing!</h1>
        <p className="text-body">
          A confirmation email has been sent to{' '}
          <span className="text-body-bold">{email}</span>. Please open it and
          click the button inside to confirm your subscription
        </p>
      </div>
      <Button type="button" label="Dismiss message" onClick={handleClick} />
    </div>
  );
}

export default ConfirmationModal;
