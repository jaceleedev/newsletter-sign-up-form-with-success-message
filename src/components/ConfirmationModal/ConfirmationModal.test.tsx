import { beforeEach, describe, expect, it, vi } from 'vitest';
import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import ConfirmationModal from './ConfirmationModal';

expect.extend(toHaveNoViolations);

vi.mock('react-dom', async () => {
  return {
    useFormStatus: vi.fn().mockReturnValue({ pending: false }),
  };
});

describe('ConfirmationModal', () => {
  beforeEach(() => {
    cleanup();
  });

  // 1. Rendering test
  it('renders correctly', () => {
    render(<ConfirmationModal email="test@example.com" />);
    const heading = screen.getByRole('heading', {
      name: 'Thanks for subscribing!',
    });
    expect(heading).toBeDefined();
  });

  // 2. Props test
  it('displays the correct email', () => {
    render(<ConfirmationModal email="test@example.com" />);
    const emailText = screen.getByText('test@example.com');
    expect(emailText).toBeDefined();
  });

  // 3. Styles test
  it('applies the correct styles', () => {
    const { container } = render(
      <ConfirmationModal email="test@example.com" />
    );
    const modal = container.firstElementChild;
    expect(modal?.className).toContain('max-w-[375px]');
    expect(modal?.className).toContain('bg-white');
  });

  // 4. State change test
  it('hides modal when dismiss button is clicked', () => {
    render(<ConfirmationModal email="test@example.com" />);
    const dismissButton = screen.getByRole('button', {
      name: 'Dismiss message',
    });
    fireEvent.click(dismissButton);
    expect(
      screen.queryByRole('heading', { name: 'Thanks for subscribing!' })
    ).toBeNull();
  });

  // 5. Behavior test
  it('calls onClick when dismiss button is clicked', () => {
    render(<ConfirmationModal email="test@example.com" />);
    const dismissButton = screen.getByRole('button', {
      name: 'Dismiss message',
    });
    fireEvent.click(dismissButton);
    expect(
      screen.queryByRole('heading', { name: 'Thanks for subscribing!' })
    ).toBeNull();
  });

  // 6. Accessibility test
  it('has no accessibility violations', async () => {
    const { container } = render(
      <ConfirmationModal email="test@example.com" />
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  // 7. Snapshot test
  it('matches snapshot', () => {
    const { asFragment } = render(
      <ConfirmationModal email="test@example.com" />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
