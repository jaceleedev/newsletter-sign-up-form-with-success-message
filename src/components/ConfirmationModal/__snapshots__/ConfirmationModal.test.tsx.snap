// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConfirmationModal > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="max-w-[375px] h-[812px] px-6 pt-[149px] pb-10 bg-white lg:max-w-[504px] lg:h-[520px] lg:px-16 lg:pt-12 lg:pb-16 lg:rounded-[36px] lg:shadow-[0_15px_60px_0_rgba(0,0,0,0.25)]"
  >
    <div
      class="mb-[263px] lg:mb-10"
    >
      <img
        alt=""
        class="mb-10"
        data-nimg="1"
        decoding="async"
        fetchpriority="high"
        height="64"
        src="/images/icon-success.svg"
        style="color: transparent;"
        width="64"
      />
      <h1
        class="text-heading mb-6"
      >
        Thanks for subscribing!
      </h1>
      <p
        class="text-body"
      >
        A confirmation email has been sent to 
        <span
          class="text-body-bold"
        >
          test@example.com
        </span>
        . Please open it and click the button inside to confirm your subscription
      </p>
    </div>
    <button
      aria-busy="false"
      class="text-body-bold text-white text-center w-full pt-[18px] pb-[14px] rounded-lg bg-dark-navy hover:bg-gradient-primary hover:shadow-primary transition-all"
      type="button"
    >
      Dismiss message
    </button>
  </div>
</DocumentFragment>
`;
